photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/ModelProcessingUtils.scala

```diff
@@ -159,53 +153,96 @@ object ModelProcessingUtils {
+          }
+        }
+        .map { innerPath =>
+
+          val name = innerPath.getName
+
+          // Load the model ID info
+          val idInfoPath = new Path(innerPath, AvroConstants.ID_INFO)
+          val Array(featureShardId) = IOUtils.readStringsFromHDFS(idInfoPath, configuration).toArray
+
+          require(
+            featureShardId != null && !featureShardId.isEmpty,
+            s"Error reading feature shard ID for coordinate '$name' in loaded model")
+
+          // Load the coefficients
+          val indexMap = featureShardIdToIndexMapLoader.get(featureShardId) match {
+
+            case Some(indexMapLoader) =>
+              indexMapLoader.indexMapForDriver()
+
+            case None =>
+              throw new IllegalArgumentException(
+                s"Missing feature shard definition for '$featureShardId' required by coordinate '$name' in loaded model")
+          }
+          val modelPath = new Path(innerPath, AvroConstants.COEFFICIENTS)
+          val glm = loadGLMFromHDFS(modelPath.toString, indexMap, sc)
+
+          (name, new FixedEffectModel(sc.broadcast(glm), featureShardId))
+        }
+
     } else {
-      Array[(String, IndexMapLoader, FixedEffectModel)]()
+      Array[(CoordinateId, FixedEffectModel)]()
     }
 
     // Load the random effect models
     val randomEffectModelInputDir = new Path(modelsDir, AvroConstants.RANDOM_EFFECT)
     val randomEffectModels = if (fs.exists(randomEffectModelInputDir)) {
 
-      fs.listStatus(randomEffectModelInputDir).map { innerFileStatus =>
+      fs
+        .listStatus(randomEffectModelInputDir)
+        .map(_.getPath)
+        .filter { innerPath =>
+          coordinatesToLoadOpt match {
+            case Some(coordinatesToLoad) => coordinatesToLoad.contains(innerPath.getName)
+            case None => true
+          }
+        }
+        .map { innerPath =>
 
-        val innerPath = innerFileStatus.getPath
-        val name = innerPath.getName
+          val name = innerPath.getName
 
-        // Load the model ID info
-        val idInfoPath = new Path(innerPath, AvroConstants.ID_INFO)
-        val Array(randomEffectType, featureShardId) = IOUtils.readStringsFromHDFS(idInfoPath, configuration).toArray
+          // Load the model ID info
+          val idInfoPath = new Path(innerPath, AvroConstants.ID_INFO)
 
-        // Load the models
-        val featureMapLoader = featureShardIdToIndexMapLoader.flatMap(_.get(featureShardId))
-        val modelsRDDInputPath = new Path(innerPath, AvroConstants.COEFFICIENTS)
-        val (modelsRDD, featureIndexLoader) =
-          loadModelsRDDFromHDFS(modelsRDDInputPath.toString, featureMapLoader, sc)
+          println(IOUtils.readStringsFromHDFS(idInfoPath, configuration).toArray.mkString(", "))
+
+          val Array(randomEffectType, featureShardId) = IOUtils.readStringsFromHDFS(idInfoPath, configuration).toArray
+
+          require(
+            featureShardId != null && !featureShardId.isEmpty,
```

Comment:
Same comment as above. I don't think `featureShardId == null` is really a reachable state here.